### PR TITLE
Add `first` and `last` method to `IndexedFrame`

### LINKED ISSUE #9600: [FEA] Series and and DataFrame.first and last
Pandas provides a `first` and `last` methods on DataFrames and Series. These methods require the index be a DatetimeIndex. They accept an offset string, DateOffset, or dateutil.relativedelta.

- [ ] [First](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.first.html): "This function can select the first few rows based on a date offset."
- [ ] [Last](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.last.html#pandas.DataFrame.last): "this function selects the last few rows based on a date offset."

These can potentially be implemented with a combination of searchsorted and slicing.

The documentation for `last` explicitly indicates the DatetimeIndex should be sorted, however neither `first` nor `last` conduct runtime checks of this requirement (see https://github.com/pandas-dev/pandas/issues/38000 ). Additionally, it's not clear that pandas unit tests cover offsets other than strings based on [these tests](https://github.com/pandas-dev/pandas/blob/dae5c5972aa539dcd5ba633e8d56735d61cfe7ae/pandas/tests/frame/methods/test_first_and_last.py).

To start, we could support only string offsets, like in the examples below.

It may be worth discussing expected behavior on the pandas repository.

For example:

```python
import pandas as pd
​
i = pd.date_range('2018-04-09', periods=4, freq='2D')
ts = pd.DataFrame({'A': [1, 2, 3, 4]}, index=i)
print(ts)
print(ts.first('3D'))
            A
2018-04-09  1
2018-04-11  2
2018-04-13  3
2018-04-15  4
            A
2018-04-09  1
2018-04-11  2
```

```python
import pandas as pd
​
i = pd.date_range('2018-04-09', periods=4, freq='2D')
ts = pd.DataFrame({'A': [1, 2, 3, 4]}, index=i)
print(ts)
print(ts.last("6D"))
            A
2018-04-09  1
2018-04-11  2
2018-04-13  3
2018-04-15  4
            A
2018-04-11  2
2018-04-13  3
2018-04-15  4
```

### PR DESCRIPTION
**Pull request by isVoid**
closes #9600 

This PR adds `first` and `last` method to `indexed_frame`.  This method only applies to `IndexedFrame` with `DatetimeIndex` and gathers the first or last rows within time range specified by `offset` argument.

### CODE REVIEW

File: python/cudf/cudf/core/indexed_frame.py [hunk 1]
```diff
@@ -758,3 +760,125 @@ def resample(
...
+
+        if len(self) == 0:
+            return self.copy()
+
+        pd_offset = pd.tseries.frequencies.to_offset(offset)
+        to_search = op(pd.Timestamp(self._index._column[idx]), pd_offset)
+        if (
+            idx == 0
+            and not isinstance(pd_offset, pd.tseries.offsets.Tick)
+            and pd_offset.is_on_offset(pd.Timestamp(self._index[0]))
```
Comment (skirui-source) [inline python/cudf/cudf/core/indexed_frame.py:783]: does `is_on_offset`  check for datetimes values only at the `end` of the offset?

**Review by skirui-source — APPROVED** [review]

File: python/cudf/cudf/core/indexed_frame.py [hunk 2]
```diff
@@ -758,3 +760,125 @@ def resample(
...
+        if not isinstance(offset, str):
+            raise NotImplementedError(
+                f"Unsupported offset type {type(offset)}."
+            )
+
+        if len(self) == 0:
+            return self.copy()
+
+        pd_offset = pd.tseries.frequencies.to_offset(offset)
+        to_search = op(pd.Timestamp(self._index._column[idx]), pd_offset)
```
[the remark this answers is not in the document]
Comment (isVoid) [reply python/cudf/cudf/core/indexed_frame.py:779]: If `first` is called, then `op` is `add` to compute the cut-off date counting from the first date in the column; if `last`, `op` is `sub` to compute the that from the last date. `to_search` is the cut-off date. Feel free to request changes if the naming/logic isn't very readable to your taste.

File: python/cudf/cudf/core/indexed_frame.py (hunk 1 above)
Comment (isVoid) [reply python/cudf/cudf/core/indexed_frame.py:783]: I think pandas `DateOffset` can be `MonthBegin` or `MonthEnd`. Depending on the string to be either `MS` or `M`.
```python
In [3]: pd.tseries.frequencies.to_offset('M')
Out[3]: <MonthEnd>

In [4]: pd.tseries.frequencies.to_offset('MS')
Out[4]: <MonthBegin>
```
I believe `is_on_offset` checks if the given datetime falls on the given offset. There isn't much documentation about this but it seems like every kind of offset has this function.

Like `Nano` would always return true:
https://github.com/pandas-dev/pandas/blob/878a0225c648cb145949f78085a8ff3f902a1c20/pandas/_libs/tslibs/offsets.pyx#L834-L835

and `BusinessDay` would check if the weekday of the given date is monday to friday
https://github.com/pandas-dev/pandas/blob/878a0225c648cb145949f78085a8ff3f902a1c20/pandas/_libs/tslibs/offsets.pyx#L1445-L1448

and for `MonthEnd` I believe it resorts to the base method:
https://github.com/pandas-dev/pandas/blob/878a0225c648cb145949f78085a8ff3f902a1c20/pandas/_libs/tslibs/offsets.pyx#L650-L659
Note that `+` and `-` on the `MonthEnd` basically sets the date to the end of that month, same apply to other offsets.

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**